File: agentpay/backend/tests/sdk/python/sdk_integration_test.py
#!/usr/bin/env python3
"""
Python SDK Integration Test
Verifies: Python SDK -> Frame API -> Payment Intent -> Policy -> Provider -> Status
"""
import sys
import os
import time
import json
import urllib.request
import urllib.error

# Add python sdk to sys.path
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), "../../../../sdk/python")))

from frame_pay.client import FrameClient, FramePaymentError

BASE_URL = "http://localhost:3001"

def assert_true(cond, msg):
    if not cond:
        print(f"❌ ASSERTION FAILED: {msg}")
        sys.exit(1)
    print(f"  ✓ {msg}")

def http_post(path, data, token=None):
    url = f"{BASE_URL}{path}"
    headers = {"Content-Type": "application/json"}
    if token:
        headers["Authorization"] = f"Bearer {token}"
    req = urllib.request.Request(url, data=json.dumps(data).encode("utf-8"), headers=headers, method="POST")
    with urllib.request.urlopen(req) as resp:
        return json.loads(resp.read().decode("utf-8"))

def main():
    print("================================================================")
    print("🐍 FRAME PYTHON SDK INTEGRATION TEST")
    print("================================================================\n")

    # 1. Health check
    req = urllib.request.Request(f"{BASE_URL}/health")
    with urllib.request.urlopen(req) as resp:
        assert_true(resp.status == 200, "Frame API health check passed")

    test_id = str(int(time.time()))

    # 2. Provision Admin User & Org via API
    print("--- 1. Provision User & Org via Frame API ---")
    reg_resp = http_post("/v1/auth/register", {
        "email": f"python_sdk_{test_id}@test.corp",
        "password": "Password123!",
        "name": "Python SDK Tester",
        "organization_name": f"Python SDK Corp {test_id}"
    })
    token = reg_resp["data"]["token"]
    org_id = reg_resp["data"]["organization"]["id"]
    assert_true(bool(token), "Admin token acquired")

    # 3. Create Agent
    print("\n--- 2. Create Agent & Credentials ---")
    agent_resp = http_post("/v1/agents", {
        "name": f"Python Procurement Bot {test_id}",
        "description": "Agent running Python SDK",
        "purpose": "Automated procurement"
    }, token=token)
    agent_id = agent_resp["data"]["id"]
    assert_true(bool(agent_id), f"Agent created with ID: {agent_id}")

    # Generate API key
    cred_resp = http_post(f"/v1/agents/{agent_id}/credentials", {}, token=token)
    agent_api_key = cred_resp["data"]["api_key"]
    assert_true(agent_api_key.startswith("frm_"), "Agent API key generated")

    # 4. Create Policy & Grant Payment Authority
    print("\n--- 3. Create Policy & Grant Payment Authority ---")
    http_post("/v1/policies", {
        "agent_id": agent_id,
        "name": f"Policy for {agent_id}",
        "transaction_limit_paise": 500000,
        "daily_limit_paise": 2000000,
        "monthly_limit_paise": 5000000,
        "approval_threshold_paise": 300000
    }, token=token)

    auth_resp = http_post("/v1/payment-authorities", {
        "agent_id": agent_id,
        "max_transaction_amount_paise": 500000, # ₹5,000
        "daily_limit_paise": 2000000,           # ₹20,000
        "monthly_limit_paise": 5000000,         # ₹50,000
        "allowed_categories": ["electronics", "office_supplies"],
        "allowed_merchants": ["tech-gear", "demo-merchant"],
        "purpose": "Python SDK hardware orders",
        "valid_until": "2030-12-31T23:59:59Z"
    }, token=token)
    auth_id = auth_resp["data"]["id"]
    assert_true(bool(auth_id), f"Authority granted: {auth_id}")

    # 5. Initialize Python FrameClient
    print("\n--- 4. Initialize Python FrameClient ---")
    client = FrameClient(api_key=agent_api_key, base_url=BASE_URL)
    assert_true(isinstance(client, FrameClient), "FrameClient instantiated successfully")

    # 6. Inspect Authorities via SDK
    print("\n--- 5. Inspect Authorities via Python SDK ---")
    authorities = client.list_authorities()
    auth_list = authorities.get("data", [])
    assert_true(len(auth_list) >= 1, "Agent authorities listed via SDK")
    target_auth = next((a for a in auth_list if a["id"] == auth_id), None)
    assert_true(target_auth is not None, f"Found target authority {auth_id} in agent's scoped authorities")

    auth_detail = client.get_authority(auth_id)
    assert_true(auth_detail.get("data", {}).get("status") == "ACTIVE", "Authority is confirmed ACTIVE")

    # 7. Execute Valid Payment (₹2,499)
    print("\n--- 6. Execute Payment (₹2,499) via Python SDK ---")
    decision = client.pay(
        amount=2499.0,
        merchant="tech-gear",
        purpose="Mechanical keyboard under ₹3,000",
        category="electronics",
        merchant_reference=f"ORD_PY_{test_id}"
    )
    assert_true(bool(decision.id), f"Payment intent created: {decision.id}")
    assert_true(int(decision.amount_paise) == 249900, "Amount converted to paise (249900)")
    assert_true(decision.merchant == "tech-gear", "Merchant matches intent")
    assert_true(decision.status in ["SUCCEEDED", "AUTHORIZED", "EXECUTING"], f"Payment executed successfully: {decision.status}")

    # 8. Check Intent Status
    print("\n--- 7. Poll Payment Intent Status via Python SDK ---")
    fetched = client.get_intent(decision.id)
    assert_true(fetched.id == decision.id, "Fetched intent matches ID")
    assert_true(fetched.status in ["SUCCEEDED", "AUTHORIZED", "EXECUTING"], f"Current status: {fetched.status}")

    # 9. Guardrails: Attempt Excessive Payment (₹12,000 > ₹5,000 limit)
    print("\n--- 8. Guardrails: Exceeding Transaction Limit ---")
    excessive = client.pay(
        amount=12000.0,
        merchant="tech-gear",
        purpose="High-end GPU purchase",
        category="electronics"
    )
    assert_true(excessive.status in ["DENIED", "PENDING_APPROVAL"], f"High amount policy enforced: {excessive.status}")

    # 10. Security: Invalid API Key Rejection
    print("\n--- 9. Security: Verify Invalid Credential Rejection ---")
    invalid_client = FrameClient(api_key="frm_test_invalid_python_key", base_url=BASE_URL)
    threw_error = False
    try:
        invalid_client.list_authorities()
    except FramePaymentError as e:
        threw_error = True
        assert_true(e.status_code == 401, f"Received expected HTTP 401: {e.code}")

    assert_true(threw_error, "Invalid credentials correctly rejected with FramePaymentError")

    print("\n================================================================")
    print("✅ FRAME PYTHON SDK INTEGRATION TEST PASSED")
    print("================================================================\n")

if __name__ == "__main__":
    main()

File: agentpay/backend/tests/sdk/typescript/sdk-integration.test.ts
// TypeScript SDK Integration Test
// Verifies: SDK -> Frame API -> Payment Intent -> Policy -> Provider -> Status
import { FrameClient, FramePaymentError } from '../../../../sdk/typescript/src/index';
import { db } from '../../../src/db';
import { ulid } from 'ulid';
import * as bcrypt from 'bcryptjs';

const BASE_URL = 'http://localhost:3001';

function assert(condition: boolean, message: string) {
  if (!condition) {
    console.error(`❌ ASSERTION FAILED: ${message}`);
    throw new Error(message);
  }
  console.log(`  ✓ ${message}`);
}

async function runTsSdkIntegrationTest() {
  console.log('================================================================');
  console.log('🧪 FRAME TYPESCRIPT SDK INTEGRATION TEST');
  console.log('================================================================\n');

  // Verify backend is up
  const healthRes = await fetch(`${BASE_URL}/health`);
  assert(healthRes.ok, 'Backend server is running on port 3001');

  // 1. Seed Org, Agent, and Authority
  const orgId = ulid();
  await db.query(`
    INSERT INTO organizations (id, name, slug, created_at, updated_at)
    VALUES ($1, 'TS SDK Test Corp', $2, NOW(), NOW())
    ON CONFLICT (id) DO NOTHING
  `, [orgId, `ts-sdk-${orgId.toLowerCase()}`]);

  const userId = ulid();
  await db.query(`
    INSERT INTO users (id, organization_id, email, password_hash, name, role, created_at, updated_at)
    VALUES ($1, $2, $3, '$2a$10$abcdefghijklmnopqrstuu', 'TS SDK Admin', 'admin', NOW(), NOW())
  `, [userId, orgId, `user_${userId.toLowerCase()}@test.corp`]);

  const rawApiKey = `frm_test_ts_${ulid().toLowerCase()}`;
  const keyHash = await bcrypt.hash(rawApiKey, 10);
  const keyPrefix = rawApiKey.substring(0, 12);
  const agentId = ulid();

  await db.query(`
    INSERT INTO agents (id, organization_id, name, description, frame_env, status, created_at, updated_at)
    VALUES ($1, $2, 'TS SDK Procurement Agent', 'Autonomous Bot using TS SDK', 'sandbox', 'active', NOW(), NOW())
  `, [agentId, orgId]);

  const credId = ulid();
  await db.query(`
    INSERT INTO agent_credentials (id, agent_id, organization_id, key_prefix, key_hash, status, created_at)
    VALUES ($1, $2, $3, $4, $5, 'active', NOW())
  `, [credId, agentId, orgId, keyPrefix, keyHash]);

  // Create Policy and Policy Version for Agent
  const policyId = ulid();
  const versionId = ulid();
  await db.query(
    `INSERT INTO policies (id, organization_id, agent_id, name, current_version_id) VALUES ($1, $2, $3, 'TS SDK Policy', $4)`,
    [policyId, orgId, agentId, versionId]
  );
  await db.query(
    `INSERT INTO policy_versions (id, policy_id, version_number, transaction_limit_paise, daily_limit_paise, monthly_limit_paise, approval_threshold_paise)
     VALUES ($1, $2, 1, 500000, 2000000, 5000000, 300000)`,
    [versionId, policyId]
  );

  // Insert an active Payment Authority with limit ₹5,000 (500000 paise)
  const authorityId = ulid();
  await db.query(`
    INSERT INTO payment_authorities (
      id, organization_id, user_id, agent_id, purpose, max_transaction_amount_paise,
      daily_limit_paise, monthly_limit_paise, allowed_categories, allowed_merchants,
      status, valid_from, valid_until, created_at, updated_at
    )
    VALUES ($1, $2, $3, $4, 'Autonomous Office Supplies Authority', 500000, 2000000, 5000000,
      $5, $6, 'ACTIVE', NOW(), NOW() + INTERVAL '30 days', NOW(), NOW())
  `, [authorityId, orgId, userId, agentId, ['electronics', 'office_supplies'], ['tech-gear', 'demo-merchant']]);

  console.log('--- 1. Initialize FrameClient with Agent API Key ---');
  const client = new FrameClient({
    apiKey: rawApiKey,
    baseUrl: BASE_URL,
  });
  assert(client instanceof FrameClient, 'FrameClient instantiated');

  console.log('\n--- 2. Inspect Authorities via SDK ---');
  const authList = await client.listAuthorities();
  assert(authList && authList.data && authList.data.length >= 1, 'Agent authorities listed successfully');
  const foundAuth = authList.data.find((a: any) => a.id === authorityId);
  assert(!!foundAuth, `Authority ${authorityId} found in agent scoped list`);

  const authDetails = await client.getAuthority(authorityId);
  assert(authDetails && authDetails.data && authDetails.data.id === authorityId, 'Authority details retrieved');
  assert(authDetails.data.status === 'ACTIVE', 'Authority is ACTIVE');

  console.log('\n--- 3. Create Valid Payment Intent via SDK (₹2,499) ---');
  const paymentDecision = await client.pay({
    amount: 2499, // ₹2,499
    currency: 'INR',
    merchant: 'tech-gear',
    purpose: 'Mechanical keyboard under ₹3,000 threshold',
    category: 'electronics',
    merchantReference: `ORD_TS_${Date.now()}`,
  });

  assert(!!paymentDecision.id, `Payment Intent created: ${paymentDecision.id}`);
  assert(Number(paymentDecision.amountPaise) === 249900, 'Amount correctly converted to paise (249900)');
  assert(paymentDecision.merchant === 'tech-gear', 'Merchant verified');
  assert(paymentDecision.requiresApproval === false, 'Within authority limit, does not require approval');

  console.log('\n--- 4. Verify Payment Status & Lifecycle via SDK ---');
  const fetchedIntent = await client.getIntent(paymentDecision.id);
  assert(fetchedIntent.id === paymentDecision.id, 'Fetched intent ID matches');
  assert(fetchedIntent.status === 'SUCCEEDED' || fetchedIntent.status === 'AUTHORIZED' || fetchedIntent.status === 'EXECUTING',
    `Intent state is valid: ${fetchedIntent.status}`);

  console.log('\n--- 5. Verify Policy Firewall Guardrails (Exceeding Limit) ---');
  // Limit is 5000, try paying 15000 (₹15,000)
  const excessiveDecision = await client.pay({
    amount: 15000,
    merchant: 'tech-gear',
    purpose: 'Bulk server rack purchase',
    category: 'electronics',
  });
  assert(
    excessiveDecision.status === 'PENDING_APPROVAL' || excessiveDecision.status === 'DENIED',
    `Excessive amount rejected or routed to approval: status=${excessiveDecision.status}`
  );

  console.log('\n--- 6. Verify Error Handling on Invalid Credentials ---');
  const invalidClient = new FrameClient({
    apiKey: 'frm_agent_invalid_secret_key_xyz',
    baseUrl: BASE_URL,
  });

  let threwAuthError = false;
  try {
    await invalidClient.listAuthorities();
  } catch (err: any) {
    threwAuthError = true;
    assert(err instanceof FramePaymentError, 'Error instance of FramePaymentError');
    assert(err.status === 401 || err.code === 'AUTHENTICATION_FAILED' || err.code === 'UNAUTHORIZED',
      `Correct authentication error code: ${err.code} (HTTP ${err.status})`);
  }
  assert(threwAuthError, 'Invalid API key correctly rejected with exception');

  console.log('\n================================================================');
  console.log('✅ FRAME TYPESCRIPT SDK INTEGRATION TEST PASSED');
  console.log('================================================================\n');
}

runTsSdkIntegrationTest()
  .then(() => process.exit(0))
  .catch((err) => {
    console.error('❌ TypeScript SDK Test failed:', err);
    process.exit(1);
  });
